internal/server/recon/recognize.go

```diff
@@ -73,6 +81,196 @@ func RecognizePlaces(
+// the query does not contain the span, returns a list with the query as the
+// only item.
+// Examples
+// query: "ab cd ef ab g", span: "ab", result: ["ab", "cd ef", "ab", "g"]
+// query: "ab cd ef g", span: "jk", result: ["ab cd ef g"]
+func splitQueryBySpan(query string, span string) []string {
+	i := 0
+	parts := []string{""}
+
+	for i < len(query) {
+		// a valid span match is one that starts at the beginning of the query or
+		// after a word separator
+		validSpanStart := true
+		if i > 0 {
+			_, validSpanStart = wordSeparators[query[i-1]]
+		}
+		if validSpanStart {
+			if strings.HasPrefix(query[i:], span) {
+				endIdx := i + len(span)
+				// a valid span match is one that ends at the end of the query or before
+				// a word separator
+				validSpanEnd := true
+				if endIdx < len(query) {
+					_, validSpanEnd = wordSeparators[query[endIdx]]
+				}
+				if validSpanEnd {
+					// valid span match is found so add the span to the list of parts
+					parts = append(parts, span)
+					// add an empty span to start the possibly next non-span part
+					parts = append(parts, "")
+					// move on to search the next part of the query after the span we just
+					// found
+					i = endIdx
+					continue
+				}
+			}
+		}
+		// this current index in the query is not part of a valid span match so just
+		// add the current character to the current non-span part
+		parts[len(parts)-1] += string(query[i])
+		// move on to search from the next index in the query
+		i += 1
+	}
+
+	// trim spaces and filter out empty parts
+	filteredParts := []string{}
+	for _, part := range parts {
+		trimmedPart := strings.TrimSpace(part)
+		if len(trimmedPart) == 0 {
+			continue
+		}
+		filteredParts = append(filteredParts, trimmedPart)
+	}
+	return filteredParts
+}
+
+// Gets the reconName for a span.
+// The logic here should correspond to the logic for processing name into
+// reconName in flume (https://source.corp.google.com/piper///depot/google3/datacommons/prophet/flume_generator/triple_helper.cc;l=168-173)
+// TODO: also clean up consecutive spaces
+func getReconName(span string) string {
+	reconName := strings.ReplaceAll(span, " ,", "")
+	reconName = strings.ReplaceAll(reconName, ",", "")
+	reconName = strings.ReplaceAll(reconName, "^", "")
+	reconName = strings.ToLower(reconName)
+	reconName = strings.TrimSpace(reconName)
+	return reconName
+}
+
+// Takes a query and returns a map of id to use for resolution to the
+// original span (part of the query) for that id.
+// A single id can have multiple spans because we process a span by removing
+// certain characters to get an id:
+// E.g., "a^b" and "ab" would both have the id "ab"
+func getId2Span(query string) map[string]map[string]struct{} {
+	id2spans := map[string]map[string]struct{}{}
+	spanTokens := strings.Split(query, " ")
+	for i := range spanTokens {
+		span := ""
+		// make n-grams from the span tokens
+		for j := i; j < int(math.Min(float64(len(spanTokens)), reconNGramLimit+1)); j++ {
```

Comment:
nit: Can this be a constant hoisted before the loops?

Reply:
done